### PR TITLE
[Audio] Schroedinger bridge model for audio processing

### PR DESCRIPTION
**Pull request by anteju**
# What does this PR do ?

This PR adds an audio processing model based on Schroedinger bridge.

**Collection**: audio

# Changelog 
Main changes
- Added SB-based audio processing model in `audio/models/enhancement.py`
- Added relevant submodules in `audio/parts/submodules/schroedinger_bridge.py`
- Added config in `examples/audio/conf` and updated `process_audio.py`

# Usage

Model can be trained using
```shell
python examples/audio/audio_to_audio_train.py --config-name schroedinger_bridge
```

Model can be evaluated using
```shell
python examples/audio/audio_to_audio_eval.py \
    model_path=${PATH_TO_NEMO_FILE} \
    dataset_manifest=${PATH_TO_DATASET_MANIFEST} \
    output_dir=${PATH_TO_OUTPUT_DIR} \
    input_key=noisy_filepath \
    target_key=clean_filepath \
    metrics=[sdr,estoi]
```

Sampler setting can be easily overridden, e.g., to change the number of sampling steps or the type of the sampling process as
```shell
python examples/audio/audio_to_audio_eval.py ... ++sampler.num_steps=${NUM_STEPS} ++sampler.process={sde or ode}
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/audio/conf/schroedinger_bridge.yaml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    target_key: clean_filepath
+    normalize_input: false # load data as is for validation, the model will normalize it for inference
+    batch_size: 4
+    shuffle: false
+    num_workers: 4
+    pin_memory: true
+
+  encoder:
+    _target_: nemo.collections.audio.modules.transforms.AudioToSpectrogram
+    fft_length: 510
```
Comment (nithinraok) [inline examples/audio/conf/schroedinger_bridge.yaml:35]: curious why fft_length an unusal number to `510`?

File: nemo/collections/audio/losses/audio.py [hunk 2]
```diff
@@ -584,3 +584,168 @@ def forward(
...
         return mse
+
+
+def calculate_mae_batch(
+    estimate: torch.Tensor,
+    target: torch.Tensor,
+    input_length: Optional[torch.Tensor] = None,
+    mask: Optional[torch.Tensor] = None,
+) -> torch.Tensor:
+    """Calculate MAE per channel.
```
Comment (nithinraok) [inline nemo/collections/audio/losses/audio.py:595]: Add full form of MAE

File: nemo/collections/audio/losses/audio.py [hunk 3]
```diff
@@ -584,3 +584,168 @@ def forward(
...
+
+    # calculate masked mean
+    mse = calculate_mean(torch.abs(err), mask=mask, dim=dim)
+
+    return mse
+
+
+class MAELoss(Loss, Typing):
+    """
+    Computes MAE loss with weighted average across channels.
```
Comment (nithinraok) [inline nemo/collections/audio/losses/audio.py:643]: full form of MAE

File: nemo/collections/audio/models/audio_to_audio.py [hunk 4]
```diff
@@ -46,7 +46,11 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
 
     def _setup_loss(self):
         """Setup loss for this model."""
-        self.loss = AudioToAudioModel.from_config_dict(self._cfg.loss)
+        if 'loss' in self._cfg:
+            self.loss = AudioToAudioModel.from_config_dict(self._cfg.loss)
+        else:
+            logging.warning('No loss function is defined in the config.')
+            self.loss = None
```
Comment (nithinraok) [inline nemo/collections/audio/models/audio_to_audio.py:53]: Do you add loss later?

File: nemo/collections/common/parts/utils.py [hunk 5]
```diff
@@ -105,3 +107,27 @@ def extend_instance(obj, mixin):
...
+def mask_sequence_tensor(tensor: torch.Tensor, lengths: torch.Tensor):
+    """
+    For tensors containing sequences, zero out out-of-bound elements given lengths of every element in the batch.
+
+    tensor: tensor of shape (B, L), (B, D, L) or (B, D1, D2, L),
+    lengths: LongTensor of shape (B,)
+    """
+    batch_size, *_, max_lengths = tensor.shape
+
+    if len(tensor.shape) == 2:
```
Comment (nithinraok) [inline nemo/collections/common/parts/utils.py:121]: Could we re-use this function here to find the mask: https://github.com/NVIDIA/NeMo/blob/a7fbf6babe6d8a2c9f53b9e72f260151833297c1/nemo/collections/asr/models/aed_multitask_models.py#L68

**Review by nithinraok — APPROVED** [review]
LGTM
